### ISSUE TITLE
Unable to create more than one subreport with charts

### ISSUE THREAD

**Issue by EnricoGsell** [FDVSolutions/DynamicJasper#16]
I am getting an exception when trying to create a report with 2 subreport which each contain a chart.
It seems to register some parameters without prefixing some unique id for each subreport.

When trying to create the report I get the following exception:
```java
ar.com.fdvs.dj.core.layout.LayoutException: Duplicate declaration of parameter: categoryAxisLabel_chart_0.
	at ar.com.fdvs.dj.core.layout.AbstractLayoutManager.applyLayout(AbstractLayoutManager.java:111)
	at ar.com.fdvs.dj.core.DynamicJasperHelper.generateJasperReport(DynamicJasperHelper.java:557)
	at ar.com.fdvs.dj.core.DynamicJasperHelper.compileOrLoadSubreports(DynamicJasperHelper.java:472)
	at ar.com.fdvs.dj.core.DynamicJasperHelper.generateJasperReport(DynamicJasperHelper.java:549)
	at ar.com.fdvs.dj.core.DynamicJasperHelper.generateJasperReport(DynamicJasperHelper.java:534)
	at jaspertest.JasperTest.createMultipleCharts(JasperTest.java:167)
	at jaspertest.JasperTest.main(JasperTest.java:75)
Caused by: ar.com.fdvs.dj.core.registration.EntitiesRegistrationException: Duplicate declaration of parameter: categoryAxisLabel_chart_0.
	at ar.com.fdvs.dj.util.LayoutUtils.registerCustomExpressionParameter(LayoutUtils.java:120)
	at ar.com.fdvs.dj.util.ExpressionUtils.createAndRegisterExpression(ExpressionUtils.java:214)
	at ar.com.fdvs.dj.domain.chart.plot.BarPlot.transform(BarPlot.java:107)
	at ar.com.fdvs.dj.domain.chart.DJChart.transform(DJChart.java:227)
	at ar.com.fdvs.dj.core.layout.AbstractLayoutManager.createChart(AbstractLayoutManager.java:1151)
	at ar.com.fdvs.dj.core.layout.AbstractLayoutManager.layoutCharts(AbstractLayoutManager.java:924)
	at ar.com.fdvs.dj.core.layout.AbstractLayoutManager.endLayout(AbstractLayoutManager.java:159)
	at ar.com.fdvs.dj.core.layout.ClassicLayoutManager.endLayout(ClassicLayoutManager.java:160)
	at ar.com.fdvs.dj.core.layout.AbstractLayoutManager.applyLayout(AbstractLayoutManager.java:106)
	... 6 more
Caused by: net.sf.jasperreports.engine.JRException: Duplicate declaration of parameter: categoryAxisLabel_chart_0.
	at net.sf.jasperreports.engine.design.JRDesignDataset.addParameter(JRDesignDataset.java:616)
	at net.sf.jasperreports.engine.design.JRDesignDataset.addParameter(JRDesignDataset.java:601)
	at net.sf.jasperreports.engine.design.JasperDesign.addParameter(JasperDesign.java:810)
	at ar.com.fdvs.dj.util.LayoutUtils.registerCustomExpressionParameter(LayoutUtils.java:118)
	... 14 more
```

To reproduce you can use the following example:
```java
 public class JasperTest {

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) 
    {
        try {
            createMultipleCharts();
            System.out.println("No error");
        } catch(Exception ex) {
            System.out.println("Expection:");
            System.out.println(ex.getMessage());
            System.out.println(org.apache.commons.lang.exception.ExceptionUtils.getStackTrace(ex));
        }
        
    }
    
    public static void createMultipleCharts() throws ClassNotFoundException, JRException
    {
        // Create a dummy style
        Style style = new Style("lala");
        style.setFont(Font.MONOSPACED_BIG);
        
        DynamicReportBuilder drb = new DynamicReportBuilder();
        
        // Create columns for the chart only reports
        AbstractColumn chartColumnDate = ColumnBuilder.getNew()
                .setColumnProperty("stringVal", String.class.getName())
                .setTitle("stringVal")
                .setWidth(Integer.valueOf(75))
                .build();
        AbstractColumn chartColumnUniqueVisits = ColumnBuilder.getNew()
                .setColumnProperty("numVal", Integer.class.getName())
                .setTitle("numVal")
                .setWidth(Integer.valueOf(75))
                .build();
        AbstractColumn chartColumnDate2 = ColumnBuilder.getNew()
                .setColumnProperty("stringVal", String.class.getName())
                .setTitle("stringVal")
                .setWidth(Integer.valueOf(75))
                .build();
        AbstractColumn chartColumnUniqueVisits2 = ColumnBuilder.getNew()
                .setColumnProperty("numVal", Integer.class.getName())
                .setTitle("numVal")
                .setWidth(Integer.valueOf(75))
                .build();
        
        // Fetch the data source
        JRDataSource mainDS = getDummyMapCollectionDataSource();
        JRMapCollectionDataSource subDS1 = getDummyMapCollectionDataSource();
        JRMapCollectionDataSource subDS2 = getDummyMapCollectionDataSource();
        
        // Add the params and the report to concatenate
        // If collection datasource is defined as param, we need to use the data (collection) instead of the datasource
        Map params = new HashMap();
        params.put("subReportData1", subDS1.getData());
        params.put("subReportData2", subDS2.getData());
        
        // Create a subreport
        List<DJChart> charts = new ArrayList<DJChart>();
        DJChart chart1_1 = createChart(chartColumnDate, chartColumnUniqueVisits, style);
        DJChart chart1_2 = createChart(chartColumnDate, chartColumnUniqueVisits, style);
        charts.add(chart1_1);
        charts.add(chart1_2);
        List<DJChart> charts2 = new ArrayList<DJChart>();
        DJChart chart2_1 = createChart(chartColumnDate2, chartColumnUniqueVisits2, style);
        charts2.add(chart2_1);
        ArrayList<AbstractColumn> chartColumns1 = new ArrayList<AbstractColumn>();
        chartColumns1.add(chartColumnDate);
        chartColumns1.add(chartColumnUniqueVisits);
        ArrayList<AbstractColumn> chartColumns2 = new ArrayList<AbstractColumn>();
        chartColumns2.add(chartColumnDate2);
        chartColumns2.add(chartColumnUniqueVisits2);
        DynamicReport chartDyn = createTableSubReport(style, chartColumns1, charts, false);
        DynamicReport chartDyn2 = createTableSubReport(style, chartColumns2, charts2, false);
        
        // Create Subreports instead of using the Dynamic Reports directly
        SubReportBuilder srb = new SubReportBuilder();
        Subreport chartSub = srb
                .setDataSource(DJConstants.DATA_SOURCE_ORIGIN_PARAMETER, DJConstants.DATA_SOURCE_TYPE_COLLECTION, "subReportData1")
                .setDynamicReport(chartDyn, new ClassicLayoutManager())
                .build();
        
        // Add the reports to the main report
        drb.addConcatenatedReport(chartSub);
        drb.addConcatenatedReport(chartDyn2, new ClassicLayoutManager(), 
                    "subReportData2", DJConstants.DATA_SOURCE_ORIGIN_PARAMETER, DJConstants.DATA_SOURCE_TYPE_COLLECTION, true);
        
        // Build the dynamic report instance
        DynamicReport dr = drb.setDefaultStyles(style, style, style, style)
                .setTitle("Awesome  Report Title")
                .setSubtitle("This report was generated at " + new Date())
                .setPrintBackgroundOnOddRows(true)
                .setUseFullPageWidth(true)
                .build();
        
        
        // Compile, fill and view
        ClassicLayoutManager lm = new ClassicLayoutManager();
        JasperReport jasperReport = DynamicJasperHelper.generateJasperReport(dr, lm, params);
        JasperPrint jasperPrint = JasperFillManager.fillReport(jasperReport, params, mainDS);
        JasperViewer.viewReport(jasperPrint);    //finally display the report report
    }
    
    private static DynamicReport createTableSubReport(Style style, List<AbstractColumn> columns, List<DJChart> charts, Boolean showTable)
    {
        DynamicReportBuilder drb = new DynamicReportBuilder();
        
        // Add the columns to the report bulder
        for (AbstractColumn column : columns) {
            drb.addColumn(column);
        }
        
        // Add the columns to the report bulder
        for (DJChart djChart : charts) {
            drb.addChart(djChart);
        }
        
        // If we want to only show the charts or also the table
        drb.setShowDetailBand(showTable);
        
        // Build the dynamic report instance
        DynamicReport dr = drb.setDefaultStyles(style, style, style, style) // TODO DO NOT CHECK IN quick hack so that no "Arial" is used, which is currently not available....
                .setTitle("Subreport")
                .setSubtitle("This subreport was generated at " + new Date())
                .setPrintBackgroundOnOddRows(true)
                .setUseFullPageWidth(true)
                .build();
        return dr;
    }
    
    private static DJChart createChart(AbstractColumn catgeory, AbstractColumn dataSerie, Style style)
    {
        // Do the chart stuff:
        DJAxisFormat categoryAxisFormat = new DJAxisFormat("category");
        categoryAxisFormat.setLabelFont(Font.ARIAL_SMALL);
        categoryAxisFormat.setLabelColor(Color.DARK_GRAY);
        categoryAxisFormat.setTickLabelFont(Font.ARIAL_SMALL);
        categoryAxisFormat.setTickLabelColor(Color.DARK_GRAY);
        categoryAxisFormat.setTickLabelMask("");
        categoryAxisFormat.setLineColor(Color.DARK_GRAY);

        DJAxisFormat valueAxisFormat = new DJAxisFormat("value");
        valueAxisFormat.setLabelFont(Font.ARIAL_SMALL);
        valueAxisFormat.setLabelColor(Color.DARK_GRAY);
        valueAxisFormat.setTickLabelFont(Font.ARIAL_SMALL);
        valueAxisFormat.setTickLabelColor(Color.DARK_GRAY);
        valueAxisFormat.setTickLabelMask("#,##0.0");
        valueAxisFormat.setLineColor(Color.DARK_GRAY);
        valueAxisFormat.setRangeMaxValueExpression(new NumberExpression(new Integer(100000)));
        
        Random rand = new Random();
        Integer max = 999999999;
        Integer min = 0;
        // nextInt is normally exclusive of the top value,
        // so add 1 to make it inclusive
        int randomNum = rand.nextInt((max - min) + 1) + min;
        
        // Create the bar chart
        DJChart djChart = new DJBarChartBuilder()
                //chart		
                .setX(20)
                .setY(10)
                .setWidth(500)
                .setHeight(250)
                .setCentered(false)
                .setBackColor(Color.LIGHT_GRAY)
                .setShowLegend(true)
                .setPosition(DJChartOptions.POSITION_HEADER)
                .setTitle("Subreport Chart Title" + rand.nextInt((max - min) + 1) + min)
                .setTitleColor(Color.DARK_GRAY)
                .setTitleFont(Font.ARIAL_BIG_BOLD)
                .setSubtitle("Subreport Chart Subtitle" + rand.nextInt((max - min) + 1) + min)
                .setSubtitleColor(Color.DARK_GRAY)
                .setSubtitleFont(Font.COURIER_NEW_BIG_BOLD)
                .setLegendColor(Color.DARK_GRAY)
                .setLegendFont(Font.COURIER_NEW_MEDIUM_BOLD)
                .setLegendBackgroundColor(Color.WHITE)
                .setLegendPosition(DJChartOptions.EDGE_BOTTOM)
                .setTitlePosition(DJChartOptions.EDGE_TOP)
                .setLineStyle(DJChartOptions.LINE_STYLE_DOTTED)
                .setLineWidth(1)
                .setLineColor(Color.DARK_GRAY)
                .setPadding(5)
                //dataset
                .setCategory((PropertyColumn) catgeory)
                .addSerie(dataSerie, "date")
                //plot
                .setShowTickMarks(true)
                .setShowTickLabels(true)
                .setShowLabels(false)
                .setCategoryAxisFormat(categoryAxisFormat)
                .setValueAxisFormat(valueAxisFormat)
                .build();
        
        return djChart;
    }
    
    private static DynamicReport createChartSubReport(AbstractColumn catgeory, AbstractColumn dataSerie, Style style)
    {
        DynamicReportBuilder drb = new DynamicReportBuilder();
        
        // Do the chart stuff:
        DJAxisFormat categoryAxisFormat = new DJAxisFormat("category");
        categoryAxisFormat.setLabelFont(Font.ARIAL_SMALL);
        categoryAxisFormat.setLabelColor(Color.DARK_GRAY);
        categoryAxisFormat.setTickLabelFont(Font.ARIAL_SMALL);
        categoryAxisFormat.setTickLabelColor(Color.DARK_GRAY);
        categoryAxisFormat.setTickLabelMask("");
        categoryAxisFormat.setLineColor(Color.DARK_GRAY);

        DJAxisFormat valueAxisFormat = new DJAxisFormat("value");
        valueAxisFormat.setLabelFont(Font.ARIAL_SMALL);
        valueAxisFormat.setLabelColor(Color.DARK_GRAY);
        valueAxisFormat.setTickLabelFont(Font.ARIAL_SMALL);
        valueAxisFormat.setTickLabelColor(Color.DARK_GRAY);
        valueAxisFormat.setTickLabelMask("#,##0.0");
        valueAxisFormat.setLineColor(Color.DARK_GRAY);
        valueAxisFormat.setRangeMaxValueExpression(new NumberExpression(new Integer(100000)));
        
        // Add the columns required for the chart
        drb.addColumn(catgeory);
        drb.addColumn(dataSerie);
        
        // Create the bar chart
        DJChart djChart = createChart(catgeory, dataSerie, style);
        drb.addChart(djChart);
        
        // Build the dynamic report instance
        DynamicReport dr = drb.setDefaultStyles(style, style, style, style) // TODO DO NOT CHECK IN quick hack so that no "Arial" is used, which is currently not available....
                .setTitle("Chart tut Subreport")
                .setSubtitle("This subreport was generated at " + new Date())
                .setShowDetailBand(false) // We dont show the columns
                .setUseFullPageWidth(true)
                .build();
        
        return dr;
    }
    
    private static AbstractColumn createColumn(String property, Class type,
            String title, int width, Style headerStyle, Style detailStyle, String fieldDescription)
            throws ColumnBuilderException {
        
        
        AbstractColumn column = ColumnBuilder.getNew()
                .setColumnProperty(property, type.getName())
                .setTitle(title)
                .setWidth(Integer.valueOf(width))
                .setStyle(detailStyle)
                .setHeaderStyle(headerStyle)
                .setFieldDescription(fieldDescription)
                .build();
        
        return column;
    }
    
    public static JRMapCollectionDataSource getDummyMapCollectionDataSource() throws JRException
    {
        // List data source
        List simpleList = new ArrayList();
        for (int i = 0; i < 10; i++) {
            Map simpleMap = new HashMap();
            simpleMap.put("stringVal", "SomeElement" + i);
            simpleMap.put("numVal", i * 10000);
            simpleList.add(simpleMap);
        }
        
        return new JRMapCollectionDataSource(simpleList);
    }
    
}
```

DynamicJasper library used: DynamicJasper-5.0.2
JasperReports library: 6.1.0
Also tested JasperReports library 6.0.0

**Comment by juanalvarezg** [comment]
Thank you very much for reporting the issue, will try to get this fixed
soon (sorry for the uncertainity)

--

*Juan Manuel Alvarez Gimenez*
JEE Architect, CTO

Office:  +54 11 5239 9899 | Cel: +54 9 11 4400 0852
Connect me on: LINKEDIN <http://ar.linkedin.com/in/jmalvarez> | SKYPE -
juan.manuel.alvarez.gimenez

*FDVSOLUTIONS | Collective Innovation*
Avenida Belgrano 1580 2º piso C1093AAA , Buenos Aires, Argentina.

2015-06-22 4:53 GMT-03:00 EnricoG <notifications@github.com>:

**Comment by juanalvarezg** [comment]
Thank you! will be available in 5.0.7 (we are now in Maven Central!)

**Comment by juanalvarezg** [comment]
Merged! thank you!
